### PR TITLE
[stable/pomerium] Scale out improvements

### PR DESCRIPTION
**Pull request by travisgroth**
#### What this PR does / why we need it:

- Switch to a Headless service to support client side load balancing in the future
- Allow independent replicaCount scaling of the different services

#### Checklist
[Place an '[x]' (no spaces) in all applicable fields. Please remove unrelated fields.]
- [x] [DCO](https://github.com/helm/charts/blob/master/CONTRIBUTING.md#sign-your-work) signed
- [x] Chart Version bumped
- [x] Variables are documented in the README.md
- [x] Title of the PR starts with chart name (e.g. `[stable/chart]`)

### CODE REVIEW

**Comment by travisgroth** [comment]
@victornoel I've done some testing that the behavior switching to headless on the 0.2.x series works, but I'd appreciate another person testing if you have the time.

**Comment by victornoel** [comment]
>
> @victornoel <https://github.com/victornoel> I've done some testing that
> the behavior switching to headless on the 0.2.x series works, but I'd
> appreciate another person testing if you have the time.
>
@travisgroth I won't be able to test until next week. I think the replica
setting change is trivial, but I'm not clear about what `ClusterIP: None`
exactly entails...

>

**Comment by travisgroth** [comment]
>but I'm not clear about what `ClusterIP: None`
exactly entails

That changes us to a headless service.  The DNS response inside the cluster will be IPs of the individual pods, which GRPC can handle correctly.  I think this can wait a few days.  Worst case, it rolls out with 0.3 where I'm more confident in the documented behavior.

**Comment by victornoel** [comment]
> > but I'm not clear about what `ClusterIP: None`
> > exactly entails
> 
> That changes us to a headless service. The DNS response inside the cluster will be IPs of the individual pods, which GRPC can handle correctly. I think this can wait a few days. Worst case, it rolls out with 0.3 where I'm more confident in the documented behavior.

@travisgroth and will this work as well for the authenticate service if it's behind an ingress for example? It means the ingress must know how to interpret DNS answers with multiple IPs… but maybe this is not a problem?
Or maybe we need different `Service` for gRPC than for HTTP?

Anyway, I think better understand why it is needed by gRPC: to be able to load-balance itself when it multiplex requests.

**Comment by travisgroth** [comment]
>will this work as well for the authenticate service if it's behind an ingress for example

I was thinking that ingress controllers bypassed the service proxy to backend pods, but I realize that may not be true after looking around.

>Or maybe we need different Service for gRPC than for HTTP?

This is likely the most correct call.

>Anyway, I think better understand why it is needed by gRPC: to be able to load-balance itself when it multiplex requests.

Yes.  gRPC will only open one connection per endpoint, and the kube proxy load balancing is connection based.  So, in the current architecture, all requests from a given Proxy will be pinned to a single authorize or authenticate pod.  While the traffic is fairly light, this makes scale out impossible.  Exposing the pods directly to gRPC allows the proxy to at least connect to all upstreams and manage its own failover.

As a further step, v0.3.0 will have client side load balancing support enabled.  This will allow Proxy to actively utilize all upstream service connections it sees via DNS and enable true scale out.  

I'm going to refactor with a separate gRPC service for authenticate.  That should be compatible with the widest number of setups.

File: stable/pomerium/templates/_helpers.tpl [hunk 1]
```diff
@@ -15,6 +15,11 @@
 {{- default (printf "%s-authenticate" .Chart.Name) .Values.authenticate.nameOverride | trunc 63 | trimSuffix "-" -}}
 {{- end -}}
 
+{{/*Expand the name of the authenticate-service.*/}}
+{{- define "pomerium.authenticate.internalName" -}}
+{{- default (printf "%s-authenticate-internal" .Chart.Name) .Values.authenticate.internalNameOverride | trunc 63 | trimSuffix "-" -}}
```
Comment (victornoel) [inline stable/pomerium/templates/_helpers.tpl]: maybe we should add also `internalNameOverride` and `internalFullnameOverride` as commented in values.yaml?

File: stable/pomerium/README.md [hunk 2]
```diff
@@ -47,48 +58,63 @@ The command removes nearly all the Kubernetes components associated with the cha
...
+## Upgrading
+
+### 2.0.0
+
+- The authenticate service is broken into two different Service objects.  One is meant for end users to interact with, and the other is for internal GRPC based communication.
+  - You must generate an SSL cert that is valid for 
+    - `[authenticate.name].[rootDomain]` | default: `authenticate.[domain]`
+    - `[authenticate.internalName].[namespace].svc.cluster.internal` | default: `authenticate-internal.default.svc.cluster.local`
+    - `[authenticate.name].[namespace].svc.cluster.internal` | default: `authenticate.default.svc.cluster.local`
+  - You may need to run `helm upgrade --force` to recreate the authorize service correctly
```
Comment (victornoel) [inline stable/pomerium/README.md]: only the authorize service or did you mean authenticate?

Comment (travisgroth) [reply stable/pomerium/README.md]: Only authorize should have a change to the ClusterIP (or at least, that's the intention).  Authenticate will get a new headless service, so it won't be impacted.

File: stable/pomerium/templates/_helpers.tpl (hunk 1 above)
Comment (travisgroth) [reply stable/pomerium/templates/_helpers.tpl]: 👍

File: stable/pomerium/README.md (hunk 2 above)
Comment (victornoel) [reply stable/pomerium/README.md]: yes, right, sorry, didn't think enough about it :)

**Comment by victornoel** [comment]
@travisgroth so this is meant to be merged when 0.3.0 is released or do you want to do it earlier?

**Comment by travisgroth** [comment]
@victornoel I've rebased on top of the pomerium v0.3.0 PR so this should be good to go now.

**Comment by victornoel** [comment]
> @victornoel I've rebased on top of the pomerium v0.3.0 PR so this should be good to go now.

Sorry, haven't had the time to test it, if you feel confident, I think this can be merged, but if you can wait a bit more, I can try a deployment with it soon.

File: stable/pomerium/templates/authenticate-internal-service.yaml [hunk 3]
```diff
@@ -0,0 +1,28 @@
...
+{{ toYaml .Values.service.annotations | indent 4 }}
+{{- end }}
+spec:
+  clusterIP: None
+  type: ClusterIP
+  ports:
+    - port: 443
+      targetPort: https
+      protocol: TCP
+      name: https
```
Comment (victornoel) [inline stable/pomerium/templates/authenticate-internal-service.yaml]: @travisgroth is that desired that there are no 9090 port on this one?

**Comment by victornoel** [comment]
@travisgroth I couldn't make it work… I regenerated the certificates (key and cert) via `generateTLS` and used them as-is, but I still got errors in the logs like this:
```
{"level":"error","addr":":443","time":"2019-09-04T17:00:39Z","message":"http: TLS handshake error from 10.242.0.164:36576: remote error: tls: bad certificate"}
```

I'm not sure what I missed, because if I use the previous values I had for those (also generated via `generateTLS` in the past), it works with chart 1.3.2…

File: stable/pomerium/templates/authenticate-internal-service.yaml (hunk 3 above)
Comment (travisgroth) [reply stable/pomerium/templates/authenticate-internal-service.yaml]: Yes - no need for the metrics port on both services.  Keeping it on the main service makes any auto discovered names a little cleaner.

**Comment by travisgroth** [comment]
>@travisgroth I couldn't make it work…

Hm.  Thanks.  Okay let me do another round of testing myself.  

Can you share any non-policy and non-secret values from your test setup?

**Comment by victornoel** [comment]
@travisgroth here it is with comments in it:
```yaml
pomerium:
  authenticate:
    idp:
      # I set it myself in my secret
      clientID: REPLACE_ME
      # I set it myself in my secret
      clientSecret: REPLACE_ME
      provider: google
      # I set that to have helm generate the env var in the deployment
      # See https://github.com/pomerium/pomerium-helm/issues/7
      serviceAccount: "true"
      url: ""
    redirectUrl: ""
    tls:
      cert: ""
      defaultIPList: []
      defaultSANList: []
      key: ""
  authorize:
    tls:
      cert: ""
      defaultIPList: []
      defaultSANList: []
      key: ""
  config:
    # I set it myself in my secret
    cookieSecret: ""
    # I generate my own config
    existingConfig: pomerium
    existingPolicy: ""
    # I normally generate my own secret but commented this to generate the certs
    # existingSecret: pomerium
    extraOpts:
      default_upstream_timeout: 5m
    # I normally set that to false, but enabled it to generate the certificates
    generateTLS: true
    policy: {}
    rootDomain: xxx.com
    # I set it myself in my secret
    sharedSecret: ""
  extraArgs: {}
  extraEnv: {}
  extraVolumes: {}
  global: {}
  image:
    pullPolicy: IfNotPresent
    repository: pomerium/pomerium
    tag: v0.3.0
  ingress:
    annotations:
      kubernetes.io/ingress.class: contour
    enabled: false
    secret:
      cert: ""
      key: ""
      name: ""
    secretName: ""
  metrics:
    enabled: false
    port: 9090
  podAnnotations: {}
  podLabels: {}
  priorityClassName: ""
  proxy:
    authenticateInternalUrl: ""
    # I generate myself the ingress for this url (as well as for the policied routes)
    authenticateServiceUrl: https://auth.cluster.aks.xxx.com
    authorizeInternalUrl: ""
    authorizeServiceUrl: ""
    tls:
      cert: ""
      defaultIPList: []
      defaultSANList: []
      key: ""
  replicaCount: 1
  resources: {}
  service:
    annotations:
      contour.heptio.com/upstream-protocol.tls: https
    externalPort: 443
    labels: {}
    type: ClusterIP
  serviceMonitor:
    enabled: false
    labels:
      release: prometheus
    namespace: ""
  tracing:
    debug: false
    enabled: false
    jaeger:
      agent_endpoint: ""
      collector_endpoint: ""
    provider: ""
```

**Comment by travisgroth** [comment]
I can't reproduce.  I've tested with:
- generateTLS on 1.3.2 -> generateTLS on 2.0.0
- 1.3.2 with existingSecret -> 2.0.0 on existing secret
- generateTLS on 2.0.0 

I don't see anything obvious in those configs that might cause a TLS issue with the changes introduced.

Which service is logging that error?  

It sounds like a pod isn't working with the right set of certs - were they all restarted correctly after updating certs?

**Comment by travisgroth** [comment]
Correction - can reproduce.  I remember why this needed to be a full major version: it changes the name of the internal authenticate service.  Without regenerating the SSL certs, it will have issues.  I must have grabbed the wrong secret when testing the upgrade path.  If you regenerate fresh from helm, it should work.

Given some [recent changes upstream](https://github.com/pomerium/pomerium/pull/290), it's probably not worth this severe of a breaking change now.  

Some of this change is still worth making, but it is significantly simpler if don't bother with AuthN.  I'll make the appropriate updates.

**Comment by victornoel** [comment]
> Correction - can reproduce. I remember why this needed to be a full major version: it changes the name of the internal authenticate service. Without regenerating the SSL certs, it will have issues. I must have grabbed the wrong secret when testing the upgrade path. If you regenerate fresh from helm, it should work.

I don't understand, I did re-generate the certs, and it is the re-generated certs that doesn't work for me. Last time I generated the certs was when `generateTLS` was introduced

My suspicion is that the generated certs are not exactly correct for my configuration and they were before, so when I generate them, something is amiss. If I had generated them at the time of 1.3.2 same problem would have happened. I will also investigate this a bit more on my side.

**Comment by victornoel** [comment]
Suprisingly enough, the SN of my current working certificate for authenticate (so with chart 1.3.2) is:
```
DNS:%!s(<nil>).xxx.com, DNS:pomerium-authenticate.proxy.svc.cluster.local
```

**Comment by victornoel** [comment]
And the new one (with this PR) that doesn't work, is:
```
DNS:authenticate.xxx.com, DNS:pomerium-authenticate-internal.proxy.svc.cluster.local
```

Honestly I'm not 100% sure I am not the one doing something wrong :)

**Comment by travisgroth** [comment]
>DNS:authenticate.xxx.com, DNS:pomerium-authenticate-internal.proxy.svc.cluster.local

Yep that's what you should get on a fresh install.  No worries, I think the simpler version will be more predictable to test :)

**Comment by travisgroth** [comment]
/hold cancel

@victornoel simplified this down to just authorize and the replica changes.  Ready for review.

**Comment by victornoel** [comment]
@travisgroth with https://github.com/pomerium/pomerium/pull/290 (for v0.4.0), authorize will be the only service exposing a grpc endpoint then, and, proxy and authenticate will only be HTTPS and thus with the present PR, the chart will be complete with respect to `ClusterIP: None` option.
Did I get everything right?

**Comment by travisgroth** [comment]
Yes, that's the plan.

**Comment by victornoel** [comment]
/lgtm

>

**Pull request merged**